crates/workspace/src/resolve.rs

```diff
@@ -0,0 +1,252 @@
+    /// the parsed [`Settings`] into the resolver.
+    pub fn load_from_paths<P: AsRef<Path>>(
+        &mut self,
+        paths: &[P],
+    ) -> Result<(), SettingsResolverError> {
+        let paths: Vec<PathBuf> = paths.iter().map(fs::normalize_path).collect();
+
+        let mut seen = FxHashSet::default();
+
+        // Load the `resolver` with `Settings` associated with each `path`
+        for path in &paths {
+            for ancestor in path.ancestors() {
+                if seen.insert(ancestor) {
+                    if let Some(toml) = find_air_toml_in_directory(ancestor) {
+                        let settings = Self::parse_settings(&toml)?;
+                        self.add(ancestor, settings);
+                        break;
+                    }
+                } else {
+                    // We already visited this ancestor, we can stop here.
+                    break;
+                }
+            }
+        }
+
+        // TODO(hierarchical): Also iterate through the directories and collect `air.toml`
+        // found nested withing the directories for hierarchical support
+
+        Ok(())
+    }
+
+    /// Parse [Settings] from a given `air.toml`
+    // TODO(hierarchical): Allow for an `extends` option in `air.toml`, which will make things
+    // more complex, but will be very useful once we support hierarchical configuration as a
+    // way of "inheriting" most top level configuration while slightly tweaking it in a nested directory.
+    fn parse_settings(toml: &Path) -> Result<Settings, ParseTomlError> {
+        let options = parse_air_toml(toml)?;
+        let settings = options.into_settings();
+        Ok(settings)
+    }
+}
+
+/// For each provided `path`, recursively search for any R files within that `path`
+/// that match our inclusion criteria
+///
+/// NOTE: Make sure that the inclusion criteria that guide `path` discovery are also
+/// consistently applied to [SettingsResolver::load_from_paths()].
+pub fn discover_r_file_paths<P: AsRef<Path>>(paths: &[P]) -> Vec<Result<PathBuf, ignore::Error>> {
+    let paths: Vec<PathBuf> = paths.iter().map(fs::normalize_path).collect();
+
+    let Some((first_path, paths)) = paths.split_first() else {
+        // No paths provided
+        return Vec::new();
+    };
+
+    // TODO: Parallel directory visitor
+    let mut builder = ignore::WalkBuilder::new(first_path);
+
+    for path in paths {
+        builder.add(path);
+    }
+
+    // TODO: Make these configurable options (possibly just one?)
+    // Right now we explicitly call them even though they are `true` by default
+    // to remind us to expose them.
+    //
+    // "This toggles, as a group, all the filters that are enabled by default"
+    // builder.standard_filters(true)
+    builder.hidden(true);
+    builder.parents(true);
+    builder.ignore(false);
+    builder.git_ignore(true);
+    builder.git_global(true);
+    builder.git_exclude(true);
```

Comment:
It's nice that we will get gitignore support out of the box so we don't format any directories marked in a .gitignore